### ISSUE TITLE
image embedding

### ISSUE THREAD

**Issue by vinnitu** [davisking/dlib#2267]
Hi

I need to get image embedding feature vector (using resnet50)

Now I do it with this code

![image](https://user-images.githubusercontent.com/432168/102687444-cdc68880-41f7-11eb-819a-11f95dde9f5b.png)

I saw your example http://www.dlib.net/dnn_introduction3_ex.cpp.html

![image](https://user-images.githubusercontent.com/432168/102687463-ef277480-41f7-11eb-83d2-85d43ab3d6a2.png)

And as I understand we can use it to embed image to feature vector, but how?
How to do it with python?

**Comment by arrufat** [comment]
Hi!
if you look at the [architecture definition for the resnet 50](http://dlib.net/resnet.h.html), you'll see something like:

``` c++
using n50  = loss_multiclass_log<fc<1000, avg_pool_everything<backbone_50<input_rgb_image>>>>;
```

The top two layers are the loss layer (which contains a softmax layer) and the fully connected layer (fc). Here, I will assume that features you're interested in are the ones that serve as the input to the fc layer. You can access them by forwarding an image through the network and then getting the output of that layer:

``` c++
net(image);
auto features = mat(net.subnet().subnet().get_output());
// equivalent to:
// auto features = mat(layer<2>(net).get_output());
```

That line will get the output and convert it into a `dlib::matrix`, that is probably easier to work with.
The shape of the the `features` matrix will be 1 row and 2048 columns.

There is no way of doing this from python, as there are no bindings for it at the moment.
It's not impossible though, you just have to write your own binding similar to [this](https://github.com/davisking/dlib/blob/20378b0819e37ff8694e872311eca16f5e711158/tools/python/src/face_recognition.cpp).

**Comment by vinnitu** [comment]
thanks

**Comment by pfeatherstone** [comment]
What do you want to use those features for? You probably want a network trained with loss_metric to extract useful features.

**Comment by arrufat** [comment]
@pfeatherstone actually, the features from a classifier that you input to the final `fc` layer carry a lot of semantic meaning, and you can compute the distances between samples in the feature space using them to perform clustering, image retrieval, etc.
The whole network made sure that they are linearly separable (hence the `fc` layer), and you can even train a SVM on top of them.

**Comment by pfeatherstone** [comment]
Okidok. I've tried using the last features of a classifier for a type of deep sort algorithm (object tracking) but didn't work very well. It was only when I used a network trained with a type of hinge loss that I got useful results. But that was my very niche use-case. Maybe using a classifier works in other circumstances. You learn something every day.
